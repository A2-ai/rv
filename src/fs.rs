use std::fs;
use std::fs::Metadata;
use std::path::Path;

use filetime::FileTime;
use walkdir::WalkDir;

const MTIME_FILE: &str = "rv.lock";

/// Copy the whole content of a folder to another folder
pub(crate) fn copy_folder(
    from: impl AsRef<Path>,
    to: impl AsRef<Path>,
) -> Result<(), std::io::Error> {
    let from = from.as_ref();
    let to = to.as_ref();

    for entry in WalkDir::new(from) {
        let entry = entry?;
        let path = entry.path();

        let relative = path.strip_prefix(from).expect("walkdir starts with root");
        let out_path = to.join(relative);

        if entry.file_type().is_dir() {
            fs::create_dir_all(&out_path)?;
            continue;
        }

        fs::copy(path, out_path)?;
    }

    Ok(())
}

fn metadata(path: impl AsRef<Path>) -> Result<Metadata, std::io::Error> {
    let path = path.as_ref();
    fs::metadata(path)
}

/// Returns the maximum mtime found in the given folder, looking at all subfolders and
/// following symlinks
/// Taken from cargo crates/cargo-util/src/paths.rs
/// We keep it simple for now and just mtime even if it causes more rebuilds than mtime + hashes
pub(crate) fn mtime_recursive(folder: impl AsRef<Path>) -> Result<FileTime, std::io::Error> {
    let meta = metadata(folder.as_ref())?;
    if !meta.is_file() {
        return Ok(FileTime::from_last_modification_time(&meta));
    }

    // TODO: filter out hidden files/folders?
    let max_mtime = WalkDir::new(folder)
        .follow_links(true)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter_map(|e| {
            if e.path_is_symlink() {
                // Use the mtime of both the symlink and its target, to
                // handle the case where the symlink is modified to a
                // different target.
                let sym_meta = match fs::symlink_metadata(e.path()) {
                    Ok(m) => m,
                    Err(err) => {
                        log::debug!(
                            "failed to determine mtime while fetching symlink metadata of {}: {}",
                            e.path().display(),
                            err
                        );
                        return None;
                    }
                };
                let sym_mtime = FileTime::from_last_modification_time(&sym_meta);
                // Walkdir follows symlinks.
                match e.metadata() {
                    Ok(target_meta) => {
                        let target_mtime = FileTime::from_last_modification_time(&target_meta);
                        Some(sym_mtime.max(target_mtime))
                    }
                    Err(err) => {
                        log::debug!(
                            "failed to determine mtime of symlink target for {}: {}",
                            e.path().display(),
                            err
                        );
                        Some(sym_mtime)
                    }
                }
            } else {
                let meta = match e.metadata() {
                    Ok(m) => m,
                    Err(err) => {
                        log::debug!(
                            "failed to determine mtime while fetching metadata of {}: {}",
                            e.path().display(),
                            err
                        );
                        return None;
                    }
                };
                Some(FileTime::from_last_modification_time(&meta))
            }
        })
        .max() // or_else handles the case where there are no files in the directory.
        .unwrap_or_else(|| FileTime::from_last_modification_time(&meta));
    Ok(max_mtime)
}
