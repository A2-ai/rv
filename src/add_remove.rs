use std::path::Path;

use std::fs;
use fs_err::write;
use toml_edit::{Array, DocumentMut, Formatted, Value};

pub fn add_dependencies(config_file: impl AsRef<Path>, deps: Vec<String>) -> Result<(), ConfigEditError> {
    config_edit(config_file, deps, add_fxn)
}

pub fn remove_dependencies(config_file: impl AsRef<Path>, deps: Vec<String>) -> Result<(), ConfigEditError> {
    config_edit(config_file, deps, remove_fxn)
}

// add and remove are almost completely the same except for the mutation of the dependency array. 
// Thus one function calling different mutation functions instead of two functions calling many shared functions
fn config_edit<F>(config_file: impl AsRef<Path>, deps: Vec<String>, edit_fxn: F) -> Result<(), ConfigEditError>
where
    F: Fn(&mut Array, Vec<String>),
{
    let config_file = config_file.as_ref();
    let config_content = fs::read_to_string(&config_file).map_err(|e| ConfigEditError {
        path: config_file.into(),
        source: ConfigEditErrorKind::Io(e),
    })?;

    let mut doc = config_content
        .parse::<DocumentMut>()
        .map_err(|e| ConfigEditError {
            path: config_file.into(),
            source: ConfigEditErrorKind::Parse(e),
        })?;

    let project_table = doc
        .get_mut("project")
        .and_then(|item| item.as_table_mut())
        .ok_or(ConfigEditError {
            path: config_file.into(),
            source: ConfigEditErrorKind::NoField("project".to_string()),
        })?;

    let project_dependencies = project_table
        .get_mut("dependencies")
        .and_then(|item| item.as_array_mut())
        .ok_or(ConfigEditError {
            path: config_file.into(),
            source: ConfigEditErrorKind::NoField("dependencies".to_string()),
        })?;

    if let Some(last) = project_dependencies.iter_mut().last() {
        last.decor_mut().set_suffix("");
    }

    edit_fxn(project_dependencies, deps);

    project_dependencies.set_trailing("\n");
    project_dependencies.set_trailing_comma(true);

    write(config_file, doc.to_string())
        .map_err(|e| ConfigEditError {
            path: config_file.into(),
            source: ConfigEditErrorKind::Io(e),
        })?;

    Ok(())
}

fn add_fxn(config_deps: &mut Array, add_deps: Vec<String>) {
    let config_dep_names = config_deps.iter().filter_map(|v| get_dependency_name(v)).collect::<Vec<_>>();
    for d in add_deps {
        if !config_dep_names.contains(&d) {
            config_deps.push(Value::String(Formatted::new(d)));
            // Couldn't format value before pushing, so adding formatting after its added
            if let Some(last) = config_deps.iter_mut().last() {
                last.decor_mut().set_prefix("\n    ");
            }        
        }
    }
}

fn remove_fxn(config_deps: &mut Array, add_deps: Vec<String>) {
    config_deps.retain(|v| {
        if let Some(name) = get_dependency_name(v) {
            !add_deps.contains(&name)
        } else {
            true
        }
    });
}

fn get_dependency_name(value: &Value) -> Option<String> {
    match value {
        Value::String(s) => Some(s.value().to_string()),
            Value::InlineTable(t) => {
                t.get("name")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string())
            },
            _ => None
    }
}

#[derive(Debug, thiserror::Error)]
#[error("Failed to edit config at `{path}`")]
#[non_exhaustive]
pub struct ConfigEditError {
    path: Box<Path>,
    source: ConfigEditErrorKind,
}

#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub enum ConfigEditErrorKind {
    Io(#[from] std::io::Error),
    Parse(#[from] toml_edit::TomlError),
    #[error("Could not find required field {0}")]
    NoField(String),
}

mod tests {
    use std::path::Path;

    use fs_err::{read_to_string, write};
    use tempfile::tempdir;

    #[test]
    fn tester() {
        let content = read_to_string("src/tests/valid_config/all_fields.toml").unwrap();
        let mut out = String::new();
        let tmp_dir = tempdir().unwrap();
        let config_file = tmp_dir.path().join("rproject.toml");
        write(&config_file, content).unwrap();
        super::add_dependencies(&config_file, vec!["test_pkg".to_string(), "dplyr".to_string()]).unwrap();
        out.push_str("===ADDED==========");
        out.push_str(&read_to_string(&config_file).unwrap());
        super::remove_dependencies(&config_file, vec!["test_pkg".to_string(), "dplyr".to_string(), "renv".to_string()]).unwrap();
        out.push_str("===REMOVED==========");
        out.push_str(&read_to_string(&config_file).unwrap());
        insta::assert_snapshot!("add_remove", out);
    }
}
