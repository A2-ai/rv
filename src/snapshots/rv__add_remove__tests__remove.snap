---
source: src/add_remove.rs
expression: doc.to_string()
---
[project]
name = "project_name"
# Can specify which version of R is required, could be used later in rv as R version manager?
r_version = "4.4.1"
description = ""
authors = [{name = "Bob", email="hello@acme.org", maintainer = true}]
license = "MIT"
keywords = []

# Are suggested deps also enforcing repository? Only used if you're making a library
suggests = []

# Order matters
repositories = [
    { alias = "cran", url = "whatever it is"},
    { alias = "mpn", url = "https://mpn.metworx.com/snapshots/stable/2020-09-20"},
]

dependencies = [
    "dplyr",
    { name = "some-package", repository = "mpn", install_suggestions = true },
    { name = "some-package", path = "../mpn", install_suggestions = true },
    { name = "some-package", git = "https://github.com/A2-ai/scicalc", tag = "v0.1.1", install_suggestions = true },
    { name = "some-package", git = "https://github.com/A2-ai/scicalc", commit = "bc50e550e432c3c620714f30dd59115801f89995", install_suggestions = true },
]

[project.urls]
homepage = ""
issues = ""
